test/libsolutil/LP.cpp

```diff
@@ -0,0 +1,314 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolutil/LP.h>
+#include <libsolutil/LinearExpression.h>
+#include <libsolutil/CommonIO.h>
+#include <libsmtutil/Sorts.h>
+#include <libsolutil/StringUtils.h>
+#include <test/Common.h>
+
+#include <boost/test/unit_test.hpp>
+
+using namespace std;
+using namespace solidity::smtutil;
+using namespace solidity::util;
+
+
+namespace solidity::util::test
+{
+
+class LPTestFramework
+{
+public:
+	LPTestFramework()
+	{
+		m_solvingState.variableNames.emplace_back("");
+	}
+
+	LinearExpression constant(rational _value)
+	{
+		return LinearExpression::factorForVariable(0, _value);
+	}
+
+	LinearExpression variable(string const& _name)
+	{
+		return LinearExpression::factorForVariable(variableIndex(_name), 1);
+	}
+
+	/// Adds the constraint "_lhs <= _rhs".
+	void addLEConstraint(LinearExpression _lhs, LinearExpression _rhs)
+	{
+		_lhs -= _rhs;
+		_lhs[0] = -_lhs[0];
```

Comment:
We are constructing the constraint `_lhs <= _rhs` here. Inside generic expressions, the constant factor is treated in the same way as variables, but constraints are of the form `a_1 * x_1 + ... + a_n * x_n <= b`, so we negate the constant term in order to move it to the other side of the equation.

Reply:
Isn't the negation of `b` happening on the previous line i.e., `_lhs -= rhs;`

Reply:
No, that takes the rhs over to the other side. We start out with:
```
a * x + b <= c * y + d
```
(``_lhs <= _rhs``)

and we want to end up with
```
a * x - c * y <= d - b
```
So you could say thayt `_lhs -= _rhs` pulls everything over to the left and then `_lhs[0] = -_lhs[0]` pulls the constant term back to the right.

Reply:
Ah I see the source of my confusion:

Linear expressions are of the form `a*X1 + b*X2 + c` i.e., factor[0] is constant
But constraints are of the form `a*X1 + b*X2 <= c` i.e., factor[0] is on the RHS not LHS

Could we please stick to one convention i.e., make constraints also `a*X1 + b*X2 + c <= 0`?

Reply:
@ekpyron also complained about that. Maybe one solution would be to use a different class for constraints, since you cannot multiply two constraints, for example.